man/drmgr.8
.\"                                                                              
.\" Copyright (C) 2020 International Business Machines                           
.\" 
.TH DRMGR "8" "June 2020" "Linux" "Linux on Power Service Tools"
.SH NAME
drmgr \- Manage DLPAR resources of a virtualized ppc64 system.


.SH SYNOPSIS
.B drmgr
.RB [ \-d
.IR detail_level ]
.RB [ \-w
.IR minutes ]
.RB [ \-C | \-\-capabilities ]
.RB [ \-h | \-\-help ]

.B drmgr
.BR \-c " {" pci " | " cpu " | " mem " | " port " | " slot " | " phb "}"

.B drmgr \-c pci \-s
.I slot_location_code
.RB { "\-i " | " \-a " | " \-r " | " \-R "}\ [ "\-I" ]

.B drmgr \-c cpu
.RB { \-a " | " \-r "} {" \-q
.I quantity
.RB "| " \-s
.RI { drc_index " | " drc_name }}

.B drmgr \-c mem
.RB { \-a " | " \-r "} {" \-q
.I quantity
.RB "| " \-s
.RI { drc_index " | " drc_name }}

.B drmgr \-c
.RB { port " | " slot " | " phb }
.RB { \-a " | " \-r "} " \-s
.RI { drc_index " | " drc_name }


.SH DESCRIPTION
.B drmgr
is a utility for managing logical and physical hot plug capable resources of a PowerVM LPAR or QEMU/KVM pSeries virtual machine via Dynamic Logical Partitioning (DLPAR) operations. Outside of hot plugging a physical PCI I/O slot the 
.B drmgr
utility is intended to be invoked remotely by a Hardware Management Console (HMC), or by a local monitoring daemon in response to a hotplug interrupt from the controlling hypervisor.

.PP
Unless directed by IBM support or performing a PCI hotplug operation use caution invoking \fBdrmgr\fR from the commandline as it can result in inconsistent state between the LPAR and management system.


.SH OPTIONS
.TP
.BI \-d " detail_level"
Set logging verbosity level.

.TP
.BI \-w " timeout"
Specify a timeout in minutes to wait on acquiring dynamic resource lock.

.TP
.B \-C, \-\-capabilities
Display DLPAR capabilities of the logical partition.

.TP
.BI \-c " drc_type"
Dynamic reconfiguration connector type to act upon from the following list:
.BR pci ", " cpu ", " mem ", " port ", " slot ", " phb "."

.SS "drmgr -c pci -s \fIslot_location_code\fR {\fB\-i\fR | \fB\-a\fR | \fB\-r\fR | \fB\-R\fR} [\fB\-I\fR]" 
Add, remove, identify, or replace a physical PCI adapter.

.TP
.BI \-s " slot_location_code"
The physical slot location code to act upon.

.TP
.B \-i
Identify the specified \fIslot_location_code\fR by turning on its visual indicator LED.

.TP
.B \-R
Interactively replace the physical PCI adapter, with \fIslot_location_code\fR, by turning on the visual indicator LED to locate the physical slot, then powering down the slot so that the current card can be removed, and finally powering the slot back up after the new adapter has been inserted into the slot.

.TP
.B \-r
Interactively remove a physical PCI adapter, from \fIslot_location_code\fR, by turning on the visual indicator LED to locate the physical slot, and then powering down the slot so that the current card can be removed.

.TP
.B \-a
Interactively add a physical PCI adapter, to \fIslot_location_code\fR, by turning on the visual indicator LED to locate the physical slot, and then powering up the slot once the new adapter has been inserted into the slot.

.TP
.B \-I
Do not perform the identify step during interactive add, remove, or replacement operations.


.SS "drmgr -c cpu \fR{\fB-s \fR{\fIdrc_index\fR | \fIdrc_name\fR} | \fB-q \fIcount\fR} {\fB-a\fR | \fB-r\fR}"
Add or remove logical cpus from the system.

.TP
.BI \-s " drc_index " | " drc_name"
The dynamic reconfiguration connector (DRC) index or name is used to identify one specific CPU on which to perform the requested DLPAR operation.

.TP
.BI \-q " quantity"
Specify a quantity of CPUs on which to perform the requested DLPAR operation.

.TP
.B \-a
Perform a DLPAR CPU(s) add operation.

.TP
.B \-r
Perform a DLPAR CPU(s) remove operation.


.SS "drmgr -c mem \fR{\fB-s \fR{\fIdrc_index\fR | \fIdrc_name\fR} | \fB-q \fIcount\fR} {\fB-a\fR | \fB-r\fR}"
Add or remove logical memory blocks (LMBs) from the system.

.TP
.BI \-s " drc_index " | " drc_name"
The dynamic reconfiguration connector (DRC) index or name is used to identify one specific logical memory block (LMB) on which to perform the requested DLPAR operation.

.TP
.BI \-q " quantity"
Specify a quantity of LMBs on which to perform the requested DLPAR operation.

.TP
.B \-a
Perform a DLPAR LMB(s) add operation.

.TP
.B \-r
Perform a DLPAR LMB(s) remove operation.


.SS "drmgr -c \fR{\fBport \fR|\fB slot \fR|\fB phb\fR} {\fB-a \fR|\fB -r\fR} \fB-s \fR{\fIdrc_index \fR|\fI drc_name\fr}"
Add or remove a logical IOA resource such as virtualized storage and network adapters, or a logical PCI Host Bus that physical PCI devices reside under including hotpluggable slots.

.TP
.BI \-s " drc_index " | " drc_name"
The dynamic reconfiguration connector (DRC) index or name is used to identify the logical resource on which to perform the associated DLPAR operation.

.TP
.B \-a
Perform a DLPAR add operation of the specified logical resource type.

.TP
.B \-r
Perform a DLPAR remove operation of the specified logical resource type.

.SH AUTHOR
.B drmgr
was written by IBM Corporation


.SH SEE ALSO
.BR lsslot "(8)"
.BR drmgr-hooks "(8)"


Changelog
powerpc-utils-1.3.10
=====================================================================

commit 8eb14d62d6b3bb14391679bd3e1b4538be9b382c
Author: Pratik R. Sampat <psampat@linux.ibm.com>
Date:   Tue Jun 8 13:10:43 2021 +0530

    ppc64_cpu: Add support to parse PAPR information for energy and frequency
    
    H_GET_EM_PARMS H_CALL was previously responsible for exporting energy,
    frequency related information in the "/proc/ppc64/lparcfg", when the
    --frequency option was invoked, however the H_GET_EM_PARMS H_CALL will
    be deprecated P10 onwards.
    The information within lparcfg was encoded and was needed to be parsed
    in order to extract the state.
    
    The new "H_GET_ENERGY_SCALE_INFO" H_CALL has the ability to export
    information in plain-text and userspace programs can simply parse this
    from a sysfs interface, with just the knowledge of the ids of the
    attributes and without the need for interpretation of the values.
    
    The new interface exports information as follows:
    /sys/firmware/papr/energy_scale_info/
       |-- <id>/
         |-- desc
         |-- value
         |-- value_desc (if exists)
       |-- <id>/
         |-- desc
         |-- value
         |-- value_desc (if exists)
    
    The commit adds capability to parse this information from the new
    interface and in case of faliure it rolls back to the lparcfg parse to
    report this information.
    The frequency information computation has been kept intact and a the
    frequencies reported from the platform is displayed in a different
    header.
    
    Sample output from the powerpc-utils tool is as follows:
    
    Power and Performance Mode: XXXX
    Idle Power Saver Status   : XXXX
    Processor Folding Status  : XXXX --> Printed if Idle power save status is supported
    
    Platform reported frequencies --> Frequencies reported from the platform's H_CALL i.e PAPR interface
    min        :    NNNN GHz
    max        :    NNNN GHz
    static     :    NNNN GHz
    
    Tool Computed frequencies
    min        :    NNNN GHz (cpu XX)
    max        :    NNNN GHz (cpu XX)
    avg        :    NNNN GHz
    
    Signed-off-by: Pratik R. Sampat <psampat@linux.ibm.com>
    Signed-off-by: Tyrel Datwyler <tyreld@linux.ibm.com>

commit 0255a7b2bc62a8b41b79e3fe446361e1278cf374
Author: Pratik R. Sampat <psampat@linux.ibm.com>
Date:   Tue Jun 8 19:22:25 2021 +0530

    ppc64_cpu: Variable maybe unitilized at the time of return
    
    Variable "rc" maybe unitilized, as rc's value is assigned within a loop
    with no guarantee the loop will execute and can return undefined values.
    
    Therefore, fix the bug by initializing the return code to zero.
    
    Fixes: 49ad3ef26883 ("This is really more of a re-write than an update")
    Signed-off-by: Pratik R. Sampat <psampat@linux.ibm.com>
    Signed-off-by: Tyrel Datwyler <tyreld@linux.ibm.com>

commit c57b66c62d96f50f3e94d17a770a4954f526a06b
Author: Sathvika Vasireddy <sv@linux.ibm.com>
Date:   Mon May 9 13:48:10 2022 +0530

    lparstat: Fix error message
    
    This patch displays proper error message in case pread() fails.
    
    Suggested-by: Paul A. Clarke <pc@us.ibm.com>
    Signed-off-by: Sathvika Vasireddy <sv@linux.ibm.com>
    Reviewed-by: Paul A. Clarke <pc@us.ibm.com>
    Signed-off-by: Tyrel Datwyler <tyreld@linux.ibm.com>

commit 89c2efa11a22bd7d2b1647b854e3bcc73a5eb9f6
Author: Sathvika Vasireddy <sv@linux.ibm.com>
Date:   Sat May 14 13:33:39 2022 +0530

    powerpc-utils: Enhance error message when lparstat -E fails on max config systems
    
    Report clear error indicating what caused lparstat -E
    option to fail on max config systems.
    
    Without this patch:
    =====================================================
     # lparstat -E
    Failed to open /sys/devices/system/cpu/cpu340/idle_purr
    =====================================================
    
    With this patch:
    =====================================================
     # lparstat -E
    Failed to open /sys/devices/system/cpu/cpu340/idle_purr: Too many open files
    =====================================================
    
    Reported-by: Pavithra Prakash <pavrampu@linux.vnet.ibm.com>
    Signed-off-by: Sathvika Vasireddy <sv@linux.ibm.com>
    Reviewed-by: Nathan Lynch <nathanl@linux.ibm.com>
    Signed-off-by: Tyrel Datwyler <tyreld@linux.ibm.com>

commit 82c89b17e00b872a18b486375a688a4fec804527
Author: Haren Myneni <haren@linux.ibm.com>
Date:   Sat May 28 18:15:35 2022 -0700

    drmgr: Add support for 'acc' command
    
    Using HMC interface, the user can change the quality of service
    credits (QoS) dynamically for NX GZIP accelerator. This interface
    invokes drmgr command which writes the new QoS credit value to
    sysfs file. So to support NX accelerator, 'acc' command is
    introduced and the drmgr is executed with following arguments.
    
    drmgr -c acc -t <accelType> -q <qos_count> -w <timeout> -d <detail level>
    
      accelType: Only GZIP accelerator type is supported now
      qos_count: New total QoS credits
    
    Signed-off-by: Haren Myneni <haren@linux.ibm.com>
    Reviewed-by: Nathan Lynch <nathanl@linux.ibm.com>
    Signed-off-by: Tyrel Datwyler <tyreld@linux.ibm.com>

commit aff2dcf4276f19aef7b499148b4d421456851e21
Author: Michal Suchanek <msuchanek@suse.de>
Date:   Mon May 2 18:44:02 2022 +0200

    hcn-init.service: Add RemainAfterExit=yes
    
    This is consistent with other services provided by powerpc-utils.
    
    Signed-off-by: Michal Suchanek <msuchanek@suse.de>
    Signed-off-by: Tyrel Datwyler <tyreld@linux.ibm.com>

commit 09846b0346273deed7b85b3f93dbd42ff9f268ef
Author: Laurent Dufour <ldufour@linux.ibm.com>
Date:   Fri Apr 15 11:13:08 2022 +0200

    man/drmgr: fix multiple typos
    
    Signed-off-by: Laurent Dufour <ldufour@linux.ibm.com>
    Signed-off-by: Tyrel Datwyler <tyreld@linux.ibm.com>

commit b46743ca68b6a06a2c82de4048d94d3d5191717f

src/drmgr/drmgr.c
/**
 * @file drmgr.c
 *
 * Copyright (C) IBM Corporation 2006
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <sys/stat.h>
#include <getopt.h>

#include "dr.h"
#include "pseries_platform.h"

#include "options.c"

#define DRMGR_ARGS	"ac:d:Iimnp:P:Qq:Rrs:w:t:hCVH"

int output_level = 1; /* default to lowest output level */

int log_fd = 0;
int action_cnt = 0;

int read_dynamic_memory_v2 = 0;

static int handle_prrn_event = 0;
static int display_usage = 0;

typedef int (cmd_func_t)(void);
typedef int (cmd_args_t)(void);
typedef void (cmd_usage_t)(char **);

int drmgr(void);
void drmgr_usage(char **);
int valid_drmgr_options(void);

struct command {
	cmd_func_t	*func;
	cmd_args_t	*validate_options;
	cmd_usage_t	*usage;
};

#define DRMGR			0
#define DRSLOT_CHRP_SLOT	1
#define DRSLOT_CHRP_PHB		2
#define DRSLOT_CHRP_PCI		3
#define DRSLOT_CHRP_MEM		4
#define DRSLOT_CHRP_HEA		5
#define DRSLOT_CHRP_CPU		6
#define DRMIG_CHRP_PMIG		7
#define DRSLOT_CHRP_PHIB	8
#define DRACC_CHRP_ACC		9

static struct command commands[] = {
	{ .func = drmgr,
	  .validate_options = valid_drmgr_options,
	  .usage = drmgr_usage,
	},
	{ .func = drslot_chrp_slot,
	  .validate_options = valid_slot_options,
	  .usage = slot_usage,
	},
	{ .func = drslot_chrp_phb,
	  .validate_options = valid_phb_options,
	  .usage = phb_usage,
	},
	{ .func = drslot_chrp_pci,
	  .validate_options = valid_pci_options,
	  .usage = pci_usage,
	},
	{ .func = drslot_chrp_mem,
	  .validate_options = valid_mem_options,
	  .usage = mem_usage,
	},
	{ .func = drslot_chrp_hea,
	  .validate_options = valid_hea_options,
	  .usage = hea_usage,
	},
	{ .func = drslot_chrp_cpu,
	  .validate_options = valid_cpu_options,
	  .usage = cpu_usage,
	},
	{ .func = drmig_chrp_pmig,
	  .validate_options = valid_pmig_options,
	  .usage = pmig_usage,
	},
	{ .func = drmig_chrp_pmig,
	  .validate_options = valid_pmig_options,
	  .usage = phib_usage,
	},
	{ .func = dracc_chrp_acc,
	  .validate_options = valid_acc_options,
	  .usage = acc_usage,
	},
};

static struct option long_options[] = {
	{"capabilities",	no_argument,	NULL, 'C'},
	{"help",		no_argument,	NULL, 'h'},
	{0,0,0,0}
};
#define MAX_USAGE_LENGTH 512
void
command_usage(struct command *command) {

	/*
	 * Display the common usage options
	 */
	fprintf(stderr, "Usage: drmgr %s",
			"[-w minutes] [-d detail_level] [-C | --capabilities] [-h | --help]\n");

	/*
	 * Now retrieve the command specific usage text
	 */
	char *buffer = zalloc(MAX_USAGE_LENGTH);
	char *pusage=buffer; /* pusage may be over written */
	command->usage(&pusage);

	fprintf(stderr, "%s\n", pusage);

	free(buffer);
}
static char *usagestr = "{-c {port | slot | phb | pci | mem | cpu | acc} | -m}\n"
	"For more information on the specific options for the various\n"
	"connector types, run drmgr -c <type> -h";

void
drmgr_usage(char **pusage)
{
	*pusage = usagestr;
}

int valid_drmgr_options(void)
{

	if (usr_drc_type == DRC_TYPE_NONE) {
		say(ERROR, "A connector type (-c) must be specified\n");
		return -1;
	}

	if (action_cnt == 0) {
		say(ERROR, "At least one action must be specified\n");
		return -1;
	}

	if (action_cnt > 1) {
		say(ERROR, "Only one action may be specified\n");
		return -1;
	}

	if ((usr_drc_count > 1) && usr_drc_name) {
		say(ERROR, "The -q and -s flags are mutually exclusive\n");
		return -1;
	}

	if (usr_timeout < 0) {
		say(ERROR, "Invalid timeout specified: %s\n", usr_timeout);
		return -1;
	}

	return 0;
}

int parse_options(int argc, char *argv[])
{
	int c;
	int option_indx;
	int option_found = 0;

	/* disable getopt error messages */
	opterr = 0;

	while ((c = getopt_long(argc, argv, DRMGR_ARGS, long_options,
				&option_indx)) != -1) {
		option_found = 1;

		switch (c) {
		    case 'a':
			usr_action = ADD;
			action_cnt++;
			break;
		    case 'c':
			usr_drc_type = to_drc_type(optarg);
			break;
		    case 'C':
			display_capabilities = 1;
			break;
		    case 'd':
			set_output_level(atoi(optarg));
			break;
		    case 'I':
			usr_slot_identification = 0;
			break;
		    case 'i':
			usr_action = IDENTIFY;
			action_cnt++;
			break;
		    case 'n':
			  /* The -n option is also used to specify a number of
			   * seconds to attempt a self-arp.  Linux ignores this
			   * for hibernation.
			   */
			usr_prompt = 0;
			break;
		    case 'p':
			usr_p_option = optarg;
			break;
		    case 'P':
			prrn_filename = optarg;
			handle_prrn_event = 1;
			break;
		    case 'q':
			usr_drc_count = strtoul(optarg, NULL, 0);
			break;
		    case 'R':
			usr_action = REPLACE;
			action_cnt++;
			break;
		    case 'r':
			usr_action = REMOVE;
			action_cnt++;
			break;
		    case 's':
			usr_drc_name = optarg;
			break;
		    case 'Q':
			usr_action = QUERY;
			action_cnt++;
			break;
		    case 'm':
			usr_action = MIGRATE;
			action_cnt++;
			break;
		    case 'w':
			usr_timeout = strtol(optarg, NULL, 10) * 60;
			break;
		    case 'h':
		    	display_usage = 1;
		    	return 0;
		    	break;
		    case 'H':
			pci_hotplug_only = 1;
			break;
		    case 't': /* target lpid (pmig, not used) */
			usr_t_option = optarg;	/* Used for Accelerator type (ex:gzip) */
			break;
		    case 'V': /* qemu virtio pci device (workaround) */
                        pci_virtio = 1;
                        break;

		    default:
			say(ERROR, "Invalid option specified '%c'\n", optopt);
			return -1;
			break;
		}
	}

	if (!option_found)
		display_usage = 1;

	return 0;
}

struct command *get_command(void)
{
	/* Unfortunately, the connector type specified doesn't always result
	 * in a 1-to-1 relationship with the resulting command to run so we
	 * have to do some extra checking to build the correct command.
	 */
	if (usr_action == MIGRATE)
		return &commands[DRMIG_CHRP_PMIG];

	if (usr_drc_name && !strncmp(usr_drc_name, "HEA", 3))
		return &commands[DRSLOT_CHRP_HEA];
	
	switch (usr_drc_type) {
	case DRC_TYPE_NONE:
		return &commands[DRMGR];
		break;
	case DRC_TYPE_PORT:
 		return &commands[DRSLOT_CHRP_HEA];
		break;
	case DRC_TYPE_SLOT:
 		return &commands[DRSLOT_CHRP_SLOT];
		break;
	case DRC_TYPE_PHB:
		return &commands[DRSLOT_CHRP_PHB];
		break;
	case DRC_TYPE_PCI:
 		return &commands[DRSLOT_CHRP_PCI];
		break;
	case DRC_TYPE_MEM:
 		return &commands[DRSLOT_CHRP_MEM];
		break;
	case DRC_TYPE_CPU:
 		return &commands[DRSLOT_CHRP_CPU];
		break;
	case DRC_TYPE_HIBERNATE:
 		usr_action = HIBERNATE;
 		return &commands[DRSLOT_CHRP_PHIB];
		break;
	case DRC_TYPE_MIGRATION:
		return &commands[DRMIG_CHRP_PMIG];
	case DRC_TYPE_ACC:
		return &commands[DRACC_CHRP_ACC];
	default:
		/* If we make it this far, the user specified an invalid
		 * connector type.
		 */
		say(ERROR, "Dynamic reconfiguration is not supported for "
		    "connector type \"%d\" on this system\n", usr_drc_type);
		break;
	}

	return &commands[DRMGR];
}

int drmgr(void) {
	say(ERROR, "Invalid command: %d\n", usr_action);
	return -1;
}

int main(int argc, char *argv[])
{
	char log_msg[DR_PATH_MAX];
	struct command *command;
	int i, rc, offset;

	switch (get_platform()) {
	case PLATFORM_UNKNOWN:
	case PLATFORM_POWERNV:
	   fprintf(stderr, "%s: is not supported on the %s platform\n",
						argv[0], platform_name);
	   exit(1);
	}

	parse_options(argc, argv);

	rc = dr_init();
	if (rc) {
		if (handle_prrn_event) {
			say(ERROR, "Failed to handle PRRN event\n");
			unlink(prrn_filename);
		}
		return rc;
	}

	if (display_capabilities) {
		print_dlpar_capabilities();
		dr_fini();
		return 0;
	}

	if (handle_prrn_event) {
		rc = handle_prrn();
		if (rc)
			say(ERROR, "Failed to handle PRRN event\n");
		unlink(prrn_filename);
		dr_fini();
		return rc;
	}

	command = get_command();

	if (display_usage) {
		command_usage(command);
		dr_fini();
		return 0;
	}

	/* Validate the options for the action we want to perform */
	rc = command->validate_options();
	if (rc) {
		dr_fini();
		return -1;
	}

	/* Validate this platform */
	if (!valid_platform("chrp")) {
		dr_fini();
		return -1;
	}

	set_timeout(usr_timeout);

	/* Log this invocation to /var/log/messages and /var/log/drmgr */
	offset = sprintf(log_msg, "drmgr: ");
	for (i = 1; i < argc; i++)
		offset += sprintf(log_msg + offset, "%s ", argv[i]);
	log_msg[offset] = '\0';
	syslog(LOG_LOCAL0 | LOG_INFO, "%s", log_msg);
	
	say(DEBUG, "%s\n", log_msg);

	/* Now, using the actual command, call out to the proper handler */
	rc = command->func();

	dr_fini();
	return rc;
}
